### PR TITLE
refactor(compiler): add `original_code` to `goog.getMsg()` options

### PR DESCRIPTION
**Pull request by dgp1130**
This links back each placeholder in a message to the original Angular template span which defines its expression. This is useful for understanding where each placeholder comes from in the context of the full message.

This is technically a feature, but I left the commit message as `refactor` to keep it out of the changelog since it's intended only to support google3.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [X] Refactoring (no functional changes, no api changes)

## What is the current behavior?

Issue Number: http://b/214103351

Angular generates `goog.getMsg()` statements for JSCompiler which do _not_ contain any options.

```html
<div i18n>Hello, {{ name }}!</div>
```

Becomes:

```typescript
const MSG_XXX = goog.getMsg(
  // Message template.
  'Hello, {$interpolation}',
  // Message parameters.
  { interpolation': '...' },
);
```

## What is the new behavior?

Angular generates `goog.getMsg()` statements for JSCompiler which _do_ contain original code in the options. This is useful for identifying specific placeholders which are otherwise named `interpolation`. See http://b/219565887 for discussion of the new `goog.getMsg()` API.

```html
<div i18n>Hello, {{ name }}!</div>
```

Becomes:

```typescript
const MSG_XXX = goog.getMsg(
  // Message template.
  'Hello, {$interpolation}',
  // Message parameters.
  { interpolation': '...' },
  // Message options.
  {
    // Original code for each placeholder.
    original_code: {
      'interpolation': '{{ name }}',
    },
  },
);
```

## Does this PR introduce a breaking change?

- [X] No

## Other information

Existing tests did *not* like this change since it required updating basically every compliance test which included i18n, so most of the PR is just fixing them. Unfortunately, the existing parser for objects in `__i18n*__` macros for tests expects an object with a maximum depth of only one (`{foo: 'bar'}`), while tests containing `original_code` go as deep as *six* (`{foo: {bar: {baz: {...}}}}`). This is compounded by the fact that Angular template evaluation also uses `{{ foo }}` which adds to the depth even though they are in string literals. I updated the object param regular expression to support a maximum of six levels of nesting in the object, which is good enough for now. Longer term we should really replace the system with a proper parser.

### CODE REVIEW

File: packages/compiler/src/render3/view/i18n/get_msg_utils.ts [hunk 1]
```diff
@@ -22,7 +22,25 @@ export function createGoogleGetMsgStatements(
   const messageString = serializeI18nMessageForGetMsg(message);
   const args = [o.literal(messageString) as o.Expression];
   if (Object.keys(params).length) {
-    args.push(mapLiteral(params, true));
+    // Message template parameters.
```
Comment (AndrewKushnir) [inline packages/compiler/src/render3/view/i18n/get_msg_utils.ts:25]: I'd propose adding a quick example of the generated code that'd be produced (so it's easier to navigate the code). Should we also rename `params` -> `placeholders`?

```suggestion
    // i18 message placeholders object, e.g. `{"interpolation": "...", "startTagSpan": "..."}`
```

File: packages/compiler/src/render3/view/i18n/get_msg_utils.ts [hunk 2]
```diff
@@ -22,7 +22,25 @@ export function createGoogleGetMsgStatements(
...
+    args.push(
+        mapLiteral(i18nFormatPlaceholderNames(params, true /* useCamelCase */), true /* quoted */));
+
+    // Message options.
+    args.push(mapLiteral({
```
Comment (AndrewKushnir) [inline packages/compiler/src/render3/view/i18n/get_msg_utils.ts:30]:
```suggestion
    // Message options object, which contains original source code for placeholders (as they are
    // present in a template, e.g. `{original_code: {"interpolation": "{{ name }}", "startTagSpan": "<span>"}}`.
    args.push(mapLiteral({
```

File: packages/compiler/src/render3/view/i18n/get_msg_utils.ts [hunk 3]
```diff
@@ -11,7 +11,7 @@ import * as o from '../../../output/output_ast';
 
 import {serializeIcuNode} from './icu_serializer';
 import {i18nMetaToJSDoc} from './meta';
-import {formatI18nPlaceholderName} from './util';
+import {formatI18nPlaceholderName, i18nFormatPlaceholderNames} from './util';
```
Comment (AndrewKushnir) [inline packages/compiler/src/render3/view/i18n/get_msg_utils.ts:14]: Those function names are confusing/inconsistent :(
It's pre-existing, so feel free to ignore (I can create a followup PR to rename them).
```suggestion
import {formatI18nPlaceholderName, formatI18nPlaceholderNamesInMap} from './util';
```

File: packages/compiler/src/render3/view/i18n/get_msg_utils.ts [hunk 4]
```diff
@@ -22,7 +22,25 @@ export function createGoogleGetMsgStatements(
...
+    // Message options.
+    args.push(mapLiteral({
+      original_code: o.literalMap(Object.keys(params).map(
+          (param) => ({
+            key: formatI18nPlaceholderName(param),
+            quoted: true,
+            value: message.placeholders[param] ?
+                o.literal(message.placeholders[param].sourceSpan.toString()) :
+                o.literal(
+                    message.placeholderToMessage[param]
```
Comment (AndrewKushnir) [inline packages/compiler/src/render3/view/i18n/get_msg_utils.ts:38]: Could you please leave a quick comment here to describe those two cases (when the second one is triggered)?

**Review by AndrewKushnir — APPROVED** [review]
@dgp1130 awesome work! 👍

File: packages/compiler/src/render3/view/i18n/get_msg_utils.ts (hunk 1 above)
Comment (dgp1130) [reply packages/compiler/src/render3/view/i18n/get_msg_utils.ts]: Renamed to `placeholderValues` and added a doc comment that shows an example Angular template and the `goog.getMsg()` statement it generates.

File: packages/compiler/src/render3/view/i18n/get_msg_utils.ts (hunk 3 above)
Comment (dgp1130) [reply packages/compiler/src/render3/view/i18n/get_msg_utils.ts]: Renamed to `formatI18nPlaceholderNamesInMap`.

File: packages/compiler/src/render3/view/i18n/get_msg_utils.ts (hunk 4 above)
Comment (dgp1130) [reply packages/compiler/src/render3/view/i18n/get_msg_utils.ts]: Added a comment for each case. Basically the first one is for typical placeholders, while the second is for ICU placeholders.

**Comment by dgp1130** [comment]
Discovered a bug in JSCompiler when specifying `original_code` so g3sync is currently failing ([b/229143878](https://buganizer.corp.google.com/issues/229143878)). Fix is already submitted, should roll out within 24 hours. I'll do a TGP after that rolls out to make sure there aren't any weird edge cases broken by this (mainly worried about some nested ICU situations).

**Comment by dgp1130** [comment]
JSCompiler fix is rolled out and standard TAP passed. I'll run a TGP tonight to make sure there are no edge cases, and hopefully we can merge tomorrow.

**Comment by dgp1130** [comment]
[Global presubmit](http://test/OCL:441838698:BASE:441886487:1649988314421:c55199c0) revealed one application which did:

```
<div i18n>
  Hello, {{ name // i18n(ph = "user
  name") }}!
</div>
```

The newline in `user\n  name` was causing some issues with `original_code`. That definitely shouldn't be allowed to being with, so I sent http://cl/442095274 to fix that app directly. I also filed https://github.com/angular/angular/issues/45653 to make this pattern a compile time error, as we shouldn't allow it in the first place.

Once the application is fixed and I get a clean presubmit, we should be good to go. I also added a test case for repeating the same placeholder multiple times in a string, which I incorrectly thought was the root cause of the google3 issue, but it is still a gap in our current i18n testing and should be included here.

**Comment by dgp1130** [comment]
[cl/442095274](http://cl/442095274) has landed and TAP is green, should be ready to submit.

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit 1fe255c76f7cbd3c21d61fa657ae4a5f2cd25e7c.

**Pull request closed without merging**